### PR TITLE
Versions and Documentation

### PR DESCRIPTION
**Pull request by oorestisime**
This PR includes PR#43 plus the commits to create a single page documentation blueprint

### CODE REVIEW

File: debsources/app/copyright/routes.py [hunk 1]
```diff
@@ -35,6 +36,16 @@ def skeleton_variables():
     lambda e: (ErrorHandler()(e, http=404), 404))
 
 
+# Before request
+@bp_copyright.before_request
+def before_request():
+    endpoints = ['license', 'file']
+    if any(endpoint in request.endpoint for endpoint in endpoints):
```
Comment (matthieucan) [inline debsources/app/copyright/routes.py]: `request.endpoint` is a string. I think what you want is
`if request.endpoint in endpoints`
Right?

File: debsources/app/doc/routes.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
+# Copyright (C) 2015  The Debdoc developers <info@doc.debian.net>.
```
Comment (matthieucan) [inline debsources/app/doc/routes.py]: I spot a too aggressive search&replace here ;)
(same below in the copyright notice)

File: debsources/app/doc/templates/doc/doc_api.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
+{#
+  Copyright (C) 2013  The Debsources developers <info@sources.debian.net>.
+  See the AUTHORS file at the top-level directory of this distribution and at
+  https://anonscm.debian.org/gitweb/?p=qa/debsources.git;a=blob;f=AUTHORS;hb=HEAD
+  License: GNU Affero General Public License, version 3 or above.
+#}
+{# copied form templates/doc_api.html #}
```
Comment (matthieucan) [inline debsources/app/doc/templates/doc/doc_api.html]: not the case anymore, right?
same remark for doc_url

File: debsources/app/patches/routes.py [hunk 4]
```diff
@@ -37,6 +37,16 @@ def skeleton_variables():
     lambda e: (ErrorHandler()(e, http=404), 404))
 
 
+# Before request
+@bp_patches.before_request
+def before_request():
+    endpoints = ['summary', 'patch_view']
```
Comment (matthieucan) [inline debsources/app/patches/routes.py]: ping for the point mentioned in irc that happens here as well (so that I don't lose track) :)

File: debsources/query.py [hunk 5]
```diff
@@ -122,9 +122,14 @@ def location_get_path_links(endpoint, path_to):
...
-        pathl.append((p, url_for(endpoint,
-                                 path_to='/'.join(path_dict[:i + 1]))))
+    if endpoint == '.versions':
+        for (i, p) in enumerate(path_dict):
+            pathl.append((p, url_for(endpoint,
+                                     packagename='/'.join(path_dict[:i + 1]))))
+    else:
+        for (i, p) in enumerate(path_dict):
+            pathl.append((p, url_for(endpoint,
+                                     path_to='/'.join(path_dict[:i + 1]))))
```
Comment (matthieucan) [inline debsources/query.py:135]: I'll need to dive deeper in the issue, because I don't have an overview of the current state, but having 2 different ways of handling this (path_to and packagename/version) according to the endpoint will lead to more and more confusion IMO.

File: debsources/tests/test_web_cp.py [hunk 6]
```diff
@@ -324,19 +324,6 @@ def test_synopsis_parsing(self):
...
         self.assertIn(synopsis, rv.data)
 
-    def test_license_404(self):
-        rv = self.app.get("/copyright/license/gnubg/1.02.000-2/foo",
-                          follow_redirects=True)
-        self.assertIn('other versions of this package are available', rv.data)
-        link = '<a href="/copyright/license/gnubg/0.90+20091206-4/">'
-        self.assertIn(link, rv.data)
-
-    def test_glob_links(self):
```
Comment (matthieucan) [inline debsources/tests/test_web_cp.py:334]: what changed with glob links? doesn't look related to the rest of the PR

**Comment by matthieucan** [comment]
See all my inline comments.
And in general: well done! That's a lot of very useful refactoring, can't wait to merge it :)

**Comment by oorestisime** [comment]
pushed 3 commits that cover the changes you suggested (last commit has 3 lines by mistake, i ll correct it while squashing it). only thing i wasn't sure what do to is with (path_to and packagename/version) review. is this blocking for this PR? i d prefer merging this and opening a bug report to keep this in mind if this is ok with you.

**Comment by oorestisime** [comment]
hey,
added another 2 commits. First one is to fix some some whitespaces or break lines i ve been finding around while testing the patch tracker, and the second just modifies some versions in the doc so they won't break in the future. 

locally tests are 100% https://paste.debian.net/363971/ but they fail on travis due to python-apt..

**Comment by matthieucan** [comment]
Well done for the doc links, and the other fixes!

Tests work fine for me as well. Dammit python-apt for Travis, I hope this will get fixed soon.

The path_to/packagename-version is the last blocker for this PR. It introduces a duplicated code pattern :/ Not sure how to solve it properly though, I admit it takes some old ways of doing stuff into the game

**Comment by matthieucan** [comment]
merged in 064285179467bb96135c8b26ccd7c12e63c67daf
